### PR TITLE
Implements consists_of

### LINKED ISSUE #3: Implement cslib_string's consist_of
The inverse of 'contains', 'consist_of' should return TRUE only when the given string consists entirely of characters in the char string.

```
*temp ws "abbbababababac"
*gosub_scene cslib_string consists_of "abc"
${cslib_ret} == true

*temp ws "abbbaba bababac"
*gosub_scene cslib_string consists_of "abc"
${cslib_ret} == false

*temp ws "     "
*gosub_scene cslib_string is_space ws " "
${cslib_ret} == true
``` 

@danspinola also suggested we could add an `is_(white?)space` alias.

### PR DESCRIPTION
**Pull request by danspinola**
```
Implements the consists_of routine in cslib_string.
Resolves #3.
```

### CODE REVIEW

**Review by CareyJWilliams — APPROVED** [review]

File: scenes/cslib_string.txt [hunk 1]
```diff
@@ -278,4 +278,39 @@
...
+*return
+
+
+*comment CONSISTS_OF
+*comment ------------------------------
+*comment Checks if the given string consists
+*comment only of characters in a given list.
+*comment ------------------------------
+*comment 		params:
+*comment 			p_string(string): the string to check for the presence of the given cheracters.
```
Comment (diegoboston) [inline scenes/cslib_string.txt]:
```suggestion
*comment 			p_string(string): the string to check for the presence of the given characters. 
```

File: scenes/cslib_string.txt [hunk 2]
```diff
@@ -278,4 +278,39 @@
...
+*comment 			p_char_list(string): a string of characters allowed in the given string
+*comment 		returns:
+*comment 			(boolean): true if p_string consists only of chars in p_char_list, else false
+*label consists_of
+*params p_string p_char_list
+*set cslib_ret true
+*if (p_char_list = "")
+	*bug cslib_string.consists_of: Character list must not be empty.
+*if (p_string = "")
+	*set cslib_ret false
```
Comment (diegoboston) [inline scenes/cslib_string.txt:300]: shouldn't this be true? 
is "" made of all capitals? all vowels? I would say yes

File: scenes/cslib_string.txt [hunk 3]
```diff
@@ -278,4 +278,39 @@
...
+*temp p_char_list_index 1
+*if (p_string_index <= p_string_len)
+	*label consists_of_char_list_loop
+	*if (p_char_list_index <= p_char_list_len)
+		*if ((p_string#p_string_index) = (p_char_list#p_char_list_index))
+			*set p_string_index +1
+			*goto consists_of_string_loop
+		*set p_char_list_index +1
+		*goto consists_of_char_list_loop
+	*set cslib_ret false
```
Comment (diegoboston) [inline scenes/cslib_string.txt:315]: you can also immediately return (after false)

File: scenes/test_string.txt [hunk 4]
```diff
@@ -103,6 +103,28 @@
...
+*temp result_expected true
+*gosub test_finish result_expected cslib_ret
+
+*gosub test_start "CONSISTS_OF #3 (neg)"
+*gosub_scene cslib_string consists_of "       cslib" " "
+*temp result_expected false
+*gosub test_finish result_expected cslib_ret
+
+*gosub test_start "CONSISTS_OF #1 (single)"
+*gosub_scene cslib_string consists_of "!" "@!$%^"
```
Comment (diegoboston) [inline scenes/test_string.txt:123]: have also a few tests on the char set (empty, one char, repeated chars etc)

**Review by diegoboston — COMMENTED** [review]
Thanks! Minor changes suggested

File: scenes/cslib_string.txt (hunk 2 above)
Comment (CareyJWilliams) [reply scenes/cslib_string.txt:300]: It's nothing: nothing consists of... nothing o.O

File: scenes/test_string.txt (hunk 4 above)
Comment (CareyJWilliams) [reply scenes/test_string.txt:123]: Unfortunately you can't test an empty char set, as that'll throw a *bug, which I haven't yet worked out how to 'catch'.

File: scenes/cslib_string.txt (hunk 3 above)
Comment (danspinola) [reply scenes/cslib_string.txt:315]: Not really. In the last iteration, when the if statement evaluates to false the routine would fall into nothing. The return statement must be on the same indentation level.

File: scenes/cslib_string.txt (hunk 2 above)
Comment (danspinola) [reply scenes/cslib_string.txt:300]: I agree. An empty string consists of nothing. My original question was when both the given string and the character list consisted of empty strings, should the routine return true? But we have decided an empty character raises an error, so that's that.

File: scenes/cslib_string.txt (hunk 3 above)
Comment (CareyJWilliams) [reply scenes/cslib_string.txt:315]: Is that the case with implicit_control_flow set to true? (which the library actually requires atm btw).

File: scenes/cslib_string.txt (hunk 1 above)
Comment (danspinola) [reply scenes/cslib_string.txt]: oops

File: scenes/cslib_string.txt (hunk 3 above)
Comment (danspinola) [reply scenes/cslib_string.txt:315]: I made it agnostic. I suspected you'd be using implicit control, but I didn't want to rely on it. Anyway, I mean the loop repeats until the walker/index is greater than the string length. So we need a return after the if.

```
LABEL CONSISTS_OF
...
REPEAT AS LONG AS THERE'S STRING TO PARSE <- this will evaluate to false in the last iteration
	...
RETURN <- we need the return here
```

Comment (diegoboston) [reply scenes/cslib_string.txt:315]: what I meant is that you could slightly reorganize the code so that, as soon as one char from the string is NOT in your set, you update the result to false and return right there - so you don't have to always scan all the way to the end

BTW, it's really hard to read a CS loop  :-)

File: scenes/test_string.txt (hunk 4 above)
Comment (diegoboston) [reply scenes/test_string.txt:123]: There is not "catch", you are right, but I have an idea for it, which may work. In startup.txt, we use *script to inject our modified Scene.prototype.bug. The original is below. We remove the "throw" and instead update an error variable (this.stats.cs_err = "Bug in function.."). The test can then check the value of the error message.

The library would still be *script free, but our testing framework could 'catch' the bugs.

I didn't test with randomtest or quicktest - there may be other details to fix.


---- BEFORE

Scene.prototype.bug = function scene_bug(message) {
  if (message) {
    message = "Bug: " + this.replaceVariables(message);
  } else {
    message = "Bug";
  }
  throw new Error(this.lineMsg() + message);
};

-- AFTER

*script   (everything on one line)
Scene.prototype.bug = function scene_bug(message) {
  if (message) {
    message = "Bug: " + this.replaceVariables(message);
  } else {
    message = "Bug";
  }
  this.stats.cs_err = message;
};

File: scenes/cslib_string.txt (hunk 3 above)
Comment (danspinola) [reply scenes/cslib_string.txt:315]: 🤔 I don't follow. The only string it follows to the end is the charset. If it gets to the end of the charset without any match, then it does return from there.

>  it's really hard to read a CS loop

I agree with reading the loops. We could add comments indicating the beginning and the end of the loop.

```
*comment BEGIN LOOP
*label consists_of_loop
...
*comment END LOOP 
```

File: scenes/test_string.txt (hunk 4 above)
Comment (CareyJWilliams) [reply scenes/test_string.txt:123]: I like this, but I'd prefer if we stored it in a mygame.js (and made the CI consume that). Long one-line script statements are just nasty.

File: scenes/cslib_string.txt (hunk 3 above)
Comment (diegoboston) [reply scenes/cslib_string.txt:315]: That's fine, as far as it works.

File: scenes/test_string.txt (hunk 4 above)
Comment (diegoboston) [reply scenes/test_string.txt:123]: Either way. If we mod *fake_choice in a similar way, we could also test cslib_menu. *fake_choice could quietly update a stat with the number of entries displayed, or even a string with all choices joined.

**Review by diegoboston — APPROVED** [review]

File: scenes/test_string.txt (hunk 4 above)
Comment (CareyJWilliams) [reply scenes/test_string.txt:123]: Tracking the bug variant here: https://github.com/ChoicescriptIDE/cslib/issues/19

I'm less sold on making a habit of this for all commands. I override a number of stuff in CSIDE and it's honestly more of a pain than it's worth. Bug is quite low-risk as it's a simple and uncommon command that's unlikely to change. Both choice commands are quite integral and are at high risk of change with future ChoiceScript updates.

If we _really_ do want to go down that route, then I suggest forking ChoiceScript and making tweaks to accommodate this (and merge from upstream periodically). But a habit of blindly overriding the prototypes completely will come back to bite us.
